File: packages/pipeline/python/embed.py
"""Embed utterances on the Apple GPU (MPS) with sentence-transformers.

Produces the same artifacts as the transformers.js embed stage: a flat
Float32 little-endian matrix row-aligned with the input jsonl, a meta json,
and a progress checkpoint that makes the run resumable. Passages are encoded
without any instruction prefix and L2-normalized, matching the js stage.

Usage:
    uv run embed.py --input ../data/utterances.jsonl --output ../data/embeddings.bin
    uv run embed.py --model BAAI/bge-base-en-v1.5 ...   # phase 5 tier
"""

import argparse
import json
import sys
import time
from pathlib import Path


def read_texts(path: Path) -> list[str]:
    texts = []
    with path.open() as f:
        for line in f:
            texts.append(json.loads(line)["text"])
    return texts


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--input", required=True, type=Path)
    parser.add_argument("--output", required=True, type=Path)
    parser.add_argument("--model", default="BAAI/bge-small-en-v1.5")
    parser.add_argument(
        "--pooling",
        choices=["mean", "cls"],
        default="mean",
        help="mean matches the transformers.js stage and the web query encoder; "
        "bge's official recipe is cls. The whole corpus and all query paths must agree.",
    )
    parser.add_argument("--batch", type=int, default=256)
    parser.add_argument("--checkpoint-every", type=int, default=40, help="batches")
    parser.add_argument("--limit", type=int, default=0, help="stop after N rows (benchmarks)")
    args = parser.parse_args()

    import numpy as np
    import torch
    from sentence_transformers import SentenceTransformer

    device = "mps" if torch.backends.mps.is_available() else "cpu"
    if device == "cpu":
        print("warning: no MPS device, running on CPU (expect ~4x slower)", flush=True)
    print(f"embedding on {device}", flush=True)
    model = SentenceTransformer(args.model, device=device)
    pooling = model[1]
    pooling.pooling_mode_cls_token = args.pooling == "cls"
    pooling.pooling_mode_mean_tokens = args.pooling == "mean"
    dim = model.get_sentence_embedding_dimension()

    texts = read_texts(args.input)
    if args.limit:
        texts = texts[: args.limit]
    total = len(texts)

    progress_path = args.output.with_suffix(".progress.json")
    meta_path = args.output.with_suffix(".meta.json")

    done = 0
    if progress_path.exists() and args.output.exists():
        done = json.loads(progress_path.read_text())["rows"]
        expected = done * dim * 4
        actual = args.output.stat().st_size
        if actual < expected:
            print(f"progress claims {done} rows but bin has fewer bytes; restart", file=sys.stderr)
            sys.exit(1)
        # Truncate any partial tail past the checkpoint.
        with args.output.open("r+b") as f:
            f.truncate(expected)
        print(f"resuming from row {done}")

    mode = "r+b" if done else "wb"
    started = time.time()
    written = done
    with args.output.open(mode) as out:
        out.seek(done * dim * 4)
        batch_index = 0
        for start in range(done, total, args.batch):
            chunk = texts[start : start + args.batch]
            vectors = model.encode(
                chunk,
                batch_size=args.batch,
                normalize_embeddings=True,
                convert_to_numpy=True,
                show_progress_bar=False,
            )
            out.write(vectors.astype("<f4").tobytes())
            written += len(chunk)
            batch_index += 1
            if batch_index % args.checkpoint_every == 0:
                out.flush()
                progress_path.write_text(json.dumps({"rows": written}))
                rate = (written - done) / (time.time() - started)
                print(f"{written}/{total} ({rate:.0f}/s)", flush=True)

    progress_path.write_text(json.dumps({"rows": written}))
    meta_path.write_text(json.dumps({"model": args.model, "dim": dim, "count": written}))
    rate = (written - done) / max(time.time() - started, 1e-9)
    print(f"done: {written} rows at {rate:.0f}/s on {device}")
    progress_path.unlink()


if __name__ == "__main__":
    main()

File: packages/pipeline/src/stages/embed.ts
import { open, readFile, rm, truncate, writeFile } from 'node:fs/promises';
import { resolve } from 'node:path';
import { pipeline } from '@huggingface/transformers';
import { EMBED_DIM, EMBED_MODEL } from '@unquote/shared';
import { DATA_DIR } from '../config.js';
import { readJsonl } from '../util/fs.js';
import { log } from '../util/log.js';
import type { Utterance } from '../types.js';

const BATCH = 64;
const CHECKPOINT_EVERY = 50; // batches
const BIN = resolve(DATA_DIR, 'embeddings.bin');
const PROGRESS = resolve(DATA_DIR, 'embeddings.progress.json');
const META = resolve(DATA_DIR, 'embeddings.meta.json');

const texts: string[] = [];
for await (const u of readJsonl<Utterance>(resolve(DATA_DIR, 'utterances.jsonl'))) {
  texts.push(u.text);
}
log.info(`embedding ${texts.length} utterances with ${EMBED_MODEL}`);

// Resume from the last checkpoint if a previous run died mid-way.
let done = 0;
try {
  done = JSON.parse(await readFile(PROGRESS, 'utf8')).rows as number;
  await truncate(BIN, done * EMBED_DIM * 4);
  log.info(`resuming at row ${done}`);
} catch {
  await rm(BIN, { force: true });
}

const embed = await pipeline('feature-extraction', EMBED_MODEL);

const file = await open(BIN, 'a');
const started = Date.now();
let sinceCheckpoint = 0;

for (let i = done; i < texts.length; i += BATCH) {
  const batch = texts.slice(i, i + BATCH);
  const output = await embed(batch, { pooling: 'mean', normalize: true });
  const data = output.data as Float32Array;
  await file.write(Buffer.from(data.buffer, data.byteOffset, data.byteLength));

  done = Math.min(i + BATCH, texts.length);
  sinceCheckpoint += 1;
  if (sinceCheckpoint >= CHECKPOINT_EVERY || done === texts.length) {
    await file.sync();
    await writeFile(PROGRESS, JSON.stringify({ rows: done }), 'utf8');
    sinceCheckpoint = 0;
  }
  if (done % 5120 < BATCH || done === texts.length) {
    const rate = done / ((Date.now() - started) / 1000);
    log.info(`embedded ${done}/${texts.length} (${Math.round(rate)}/s)`);
  }
}

await file.close();
await writeFile(
  META,
  JSON.stringify({ model: EMBED_MODEL, dim: EMBED_DIM, count: texts.length }, null, 2),
  'utf8',
);
await rm(PROGRESS, { force: true });
log.info(`embeddings complete: ${texts.length} rows, ${EMBED_DIM} dims`);
